Integrate rustfmt

valentinewallace opened the pull request:
Closes #410. @jkczyz let me know any initial feedback!

~~Some notes:~~
~~1) I stuck with only "stable" features of `rustfmt`, of which there are only 14~~
~~2) I went with 80-character lines because I like being able to have many vim panes side-by-side, but it looks a bit weird at times. 100 might be better.~~
~~3) Travis will fail if we run check the formatting w/ multiple rust versions, so I chose `stable` for the `lightning` crate and `1.34.2` for the `fuzz` crate.~~

*Problem: the build fails on `1.22.0` -- `rustfmt` only works on `1.24.0` and beyond. Are there any plans to deprecate support for 1.22.0, and/or could we possibly bump the minimum supported version of RL to 1.24.0?*

TODO:
[ ] add a bunch of `#[rustfmt::skip]`s to places that benefit from being aligned in a specific way

TheBlueMatt commented:
Nice. 80 char-wide is really really awkward for some of our long-names types and for many of our many-parameter functions. If anything I'd vote for something more like 150, even.
How stable is rustfmt across recent versions? Does 1.40 produce something different from 1.39 from 1.38? If it does, I'm not sure its really a candidate for everyday use, no?

valentinewallace commented:
> How stable is rustfmt across recent versions? Does 1.40 produce something different from 1.39 from 1.38? 

1.40 the same as 1.39, but slightly different from 1.38. So, it's not the most stable. But, can't we all just decide on a canonical version up-front?

>Nice. 80 char-wide is really really awkward for some of our long-names types

Hm, true. Upped to 120, and it seems much improved, lmk if you still prefer >=150.

TheBlueMatt commented on lightning-net-tokio/src/lib.rs:
```diff
@@ -42,65 +42,96 @@ pub struct Connection {
-				Ok(pause_read) => {
-					if pause_read {
+		tokio::spawn(
+			reader
```
This is super awkward. Anything we can do with the defaults so it doesn't make a mess like this?

valentinewallace replied:
I think changing this default might help: https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#function-calls

TheBlueMatt replied:
Sure, lets try that.

valentinewallace replied:
Should be better with this new change!

TheBlueMatt commented on lightning/src/ln/chanmon_update_fail_tests.rs:
```diff
@@ -65,7 +89,16 @@ fn do_test_simple_monitor_temporary_update_fail(disconnect: bool) {
 	if disconnect {
 		nodes[0].node.peer_disconnected(&nodes[1].node.get_our_node_id(), false);
 		nodes[1].node.peer_disconnected(&nodes[0].node.get_our_node_id(), false);
-		reconnect_nodes(&nodes[0], &nodes[1], (true, true), (0, 0), (0, 0), (0, 0), (0, 0), (false, false));
```
This also feels a bit awkward, no? Can we make it just put a line per *thing* if the total line is too long, not just all the time?

valentinewallace replied:
I'm used to this look from previous formatters but fine with trying something different. 

>Can we make it just put a line per thing if the total line is too long, not just all the time? 

IIUC that is what it is: https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#tall-default
We could change it to this instead: https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#compressed

TheBlueMatt replied:
Hmm, my question was more about when rustfmt decides to put it on new lines...I kinda assumed it was "too long" by char count, but looks like this one shouldn't hit that metric?

valentinewallace replied:
I think the formatter fudges the character count a bit such that this line is barely over the limit: https://i.imgur.com/EWGbOBk.png 
We could increase the character count again from 120 to 150. that does seem a bit long to me but not a hill i'm willing to die on :stuck_out_tongue:

TheBlueMatt replied:
Hmm...So my smallest terminal is 140-wide, though I presume I'm a minority here. Maybe bump again and see where we end up.

jkczyz replied:
Functions with large number of arguments may signal a need to refactor the code. Here, the arguments consisting of long sequences of literals aren't very meaningful to the reader.

Generally, I'd prefer if lines aren't longer than 100-120 characters. @valentinewallace Does `fn_args_layout` apply to function definition or calls (or both)?

valentinewallace replied:
@jkczyz It appears to be both. I also prefer keeping line lengths to 100-120.

valentinewallace replied:
@jkczyz looks like it applies to both. I also prefer 100-120 character max lines.

valentinewallace commented on fuzz/src/chanmon_consistency.rs:
```diff
@@ -188,13 +182,12 @@ pub fn do_test(data: &[u8]) {
-			monitor)
-		} }
+			(ChannelManager::new(Network::Bitcoin, fee_est.clone(), monitor.clone(), broadcast.clone(), Arc::clone(&logger), keys_manager.clone(), config, 0).unwrap(), monitor)
+			}};
```
Done! good idea.

valentinewallace replied:
Oh, note that adding unstable features means we can only run the formatter on the rust nightly channel though, which requires adding another travis build and also potentially extra code churn over time. I'm fine with all that, but worth pointing out.

valentinewallace commented on fuzz/src/chanmon_consistency.rs:
```diff
@@ -494,86 +474,100 @@ pub fn do_test(data: &[u8]) {
+							} else {
+								false
+							}
+						}
```
Hm, new issue, maybe?

TheBlueMatt commented:
My biggest concern would be code churn. The only thing worse than no common format is a common format that requires regular changes and PRs with conflicts.

jkczyz commented on lightning/src/chain/keysinterface.rs:
```diff
@@ -177,15 +177,29 @@ pub struct InMemoryChannelKeys {
+	}
+	fn commitment_seed(&self) -> &[u8; 32] {
+		&self.commitment_seed
+	}
```
Could you move `fn_single_line` to a separate commit for now? Turns out it doesn't just allow single-line functions but rather it forces them for all functions of single expressions since there is a large `max_width`.  The churn is making it difficult to review the other changes.

valentinewallace replied:
oops, fixed!

valentinewallace commented:
> My biggest concern would be code churn.
@TheBlueMatt is this concern because I switched to using the nightly version?

TheBlueMatt commented:
Right, as you indicated it might cause churn, but also any inconsistencies in rustfmt would as well - you said 1.38 has different results from 1.39/1.40 which is also concerning. I presume its only getting better so I'm not objecting to it in practice, only noting that any decisions we make wrt policy should prioritize stability of output over most other issues, and that probably means not using unstable features until they stabilize.

valentinewallace commented:
Gotcha, that'll necessitate the removal of `indent_style = "Visual"` which solved the awkward parts you mentioned in [this comment](https://github.com/rust-bitcoin/rust-lightning/pull/425#discussion_r360998427). 

@jkczyz objections to prioritizing stability over the usage of `match_block_trailing_comma`, `format_macro_bodies` and `fn_single_line`?

valentinewallace commented:
>I'd prefer stability, too, given my assessment below. But without format_macro_bodies disabled, we're going to get some bad formatting (misaligned braces).

@jkczyz there's about ~100 macros (at least according to `rg "macro_rules" | wc -l`). I could add `#[rustfmt::skip]` to all of them as a stopgap, which I'd sort of prefer over switching from tabs to spaces.

jkczyz commented:
> > I'd prefer stability, too, given my assessment below. But without format_macro_bodies disabled, we're going to get some bad formatting (misaligned braces).
> 
> @jkczyz there's about ~100 macros (at least according to `rg "macro_rules" | wc -l`). I could add `#[rustfmt::skip]` to all of them as a stopgap, which I'd sort of prefer over switching from tabs to spaces.

Good call! 👍

TheBlueMatt commented on fuzz/src/bin/gen_target.sh:
```diff
@@ -14,29 +14,26 @@ GEN_TEST router
 
 GEN_TEST msg_accept_channel msg_targets::
```
Why is a .sh file being formatted?

valentinewallace replied:
I had to edit some of these files so the autogenerated code matches what Rustfmt produces. Alternately, I could have these .sh files add skips to the files they generate.

Making rustfmt ignore a file is an unstable feature, sadly.

TheBlueMatt replied:
Right, but why reorder things within the file? That shouldnt effect the generated formats?

valentinewallace replied:
ah true, fixed

TheBlueMatt replied:
Next question, why is a .h file being formatted by rustfmt?

valentinewallace replied:
Uh that's a good point. Updated!

TheBlueMatt commented on fuzz/src/full_stack.rs:
```diff
@@ -237,51 +260,136 @@ impl KeysInterface for KeyProvider {
 		let ctr = self.counter.fetch_add(1, Ordering::Relaxed);
-		[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-		(ctr >> 8*7) as u8, (ctr >> 8*6) as u8, (ctr >> 8*5) as u8, (ctr >> 8*4) as u8, (ctr >> 8*3) as u8, (ctr >> 8*2) as u8, (ctr >> 8*1) as u8, 14, (ctr >> 8*0) as u8]
+		[
```
This probably needs a skip, right?

valentinewallace commented:
>my real question is how stable is rustfmt across rust versions?

Here's the stability historically (pretty stable IMO):
1.40 == 1.39 !=(1 line of difference) 1.38 == 1.37 == 1.36 == 1.35 == 1.34

TheBlueMatt commented on .travis.yml:
```diff
@@ -2,7 +2,7 @@ language: rust
   - stable
   - beta
-  - 1.22.0
+  - 1.24.0
```
What change required this?

valentinewallace replied:
rustfmt only supported on 1.24.0 and beyond :/

TheBlueMatt replied:
Right, but we can just...not format on travis with 1.22? We are only going to require formatting against stable/1.34 or whatever, anyway, no?

valentinewallace replied:
Ah right, rustfmt doesn't need to support 1.22 for code that uses it to build on 1.22 :woman_facepalming: fixed!

TheBlueMatt commented on fuzz/rustfmt.toml:
```diff
@@ -0,0 +1,4 @@
+hard_tabs = true # use tab characters for indentation, spaces for alignment
+use_field_init_shorthand = true
+max_width = 120
+reorder_imports = false
```
Why not?

TheBlueMatt commented on fuzz/src/chanmon_consistency.rs:
```diff
@@ -135,52 +153,434 @@ impl KeysInterface for KeyProvider {
+			0,
+			0,
+			0,
+			0,
```
Gah. Can we ignore this and the next few places?

TheBlueMatt commented on fuzz/src/full_stack.rs:
```diff
@@ -867,6 +1063,7 @@ mod tests {
 		assert_eq!(log_entries.get(&("lightning::ln::peer_handler".to_string(), "Handling UpdateHTLCs event in peer_handler for node 030000000000000000000000000000000000000000000000000000000000000000 with 0 adds, 0 fulfills, 1 fails for channel 3d00000000000000000000000000000000000000000000000000000000000000".to_string())), Some(&2)); // 9
-		assert_eq!(log_entries.get(&("lightning::ln::channelmonitor".to_string(), "Input spending remote commitment tx (00000000000000000000000000000000000000000000000000000000000000fd:0) in 0000000000000000000000000000000000000000000000000000000000000044 resolves outbound HTLC with payment hash ff00000000000000000000000000000000000000000000000000000000000000 with timeout".to_string())), Some(&1)); //  ...
+		assert_eq!(log_entries.get(&("lightning::ln::channelmonitor".to_string(), "Input spending remote commitment tx (00000000000000000000000000000000000000000000000000000000000000fd:0) in 0000000000000000000000000000000000000000000000000000000000000044 resolves outbound HTLC with payment hash ff00000000000000000000000000000000000000000000000000000000000000 with timeout".to_string())), Some(&1));
+		// 10
```
Maybe just move all the comments at the end of these lines above the line?

TheBlueMatt commented on fuzz/src/msg_targets/gen_target.sh:
```diff
@@ -7,35 +7,33 @@ GEN_TEST() {
 
-echo "mod utils;" > mod.rs
+echo "mod utils;
+" > mod.rs
```
I thought we were going to stop formatting .sh files?

valentinewallace replied:
This file I switched up so that the code it generates (in mod.rs) corresponds to rustfmt

TheBlueMatt commented on lightning/src/ln/chanmon_update_fail_tests.rs:
```diff
@@ -24,17 +24,29 @@ fn test_simple_monitor_permanent_update_fail() {
 
-	let route = nodes[0].router.get_route(&nodes[1].node.get_our_node_id(), None, &Vec::new(), 1000000, TEST_FINAL_CLTV).unwrap();
+	let route = nodes[0]
+		.router
```
This is unreadable to me.

valentinewallace replied:
improved?

valentinewallace commented:
>Also, as a reviewability note, can we split it into one/two/three commits that just add exceptions, then one commit that just does a rustfmt pass, then one commit that adds it to travis etc?

I changed it, but would be fine to make more changes, if it makes it more reviewable.

D4nte commented:
Can I help with this?

TheBlueMatt commented:
Right, probably time to rerun this and check to see how stable rustfmt is across rust versions. Probably makes sense to take this after we land the few big-uns for 0.0.11.

valentinewallace commented:
> > my real question is how stable is rustfmt across rust versions?
> 
> Here's the stability historically (pretty stable IMO):
> 1.40 == 1.39 !=(1 line of difference) 1.38 == 1.37 == 1.36 == 1.35 == 1.34

Ran it on the freshest 2 versions, and `cargo +1.42.0 fmt` == 1.41.0 == 1.40.0.

TheBlueMatt commented:
Cool. Wanna rebase this on master?

valentinewallace commented:
@TheBlueMatt Update:
With these two lines added:
```
use_small_heuristics = "Max"
fn_args_layout = "Compressed"
```
the stability is this:
1.42.0 == 1.41.0 == 1.40.0 == 1.39.0 != 1.38.0 (2 lines of difference) == 1.37.0 

And 1.36.0 and below are broken due to `fn_args_layout`

If I remove `fn_args_layout` and keep `use_small_heuristics`:
1.42.0 == 1.41.0 == 1.40.0 == 1.39.0 != 1.38.0 (2 lines of difference) == 1.37.0 == 1.36.0 == 1.35.0 != 1.34.0 (diff pictured here: https://i.imgur.com/cUkAZXJ.png)

TheBlueMatt commented:
Opened an issue upstream for the somewhat absurdly tall code this seems to generate. https://github.com/rust-lang/rustfmt/issues/4146

TheBlueMatt commented:
Until rustfmt upstream gets its act together and generates sane code, lets close this. We still have an issue to track it.
